source/framework/core/src/TRestHits.cxx

```diff
@@ -541,109 +759,204 @@ void TRestHits::GetBoundaries(std::vector<double>& dist, double& max, double& mi
+///
 Double_t TRestHits::GetGaussSigmaX() {
     Double_t gausSigmaX = 0;
     Int_t nHits = GetNumberOfHits();
-    vector<Double_t> x(nHits), y(nHits), ex(nHits), ey(nHits);
-    if (nHits <= 3) {
+    if (nHits <= 0) {
         gausSigmaX = 0;
     } else {
-        for (int n = 0; n < GetNumberOfHits(); n++) {
-            x[n] = fX[n];
-            y[n] = fEnergy[n];
-            ex[n] = 0;
-            if (y[n] != 0) {
-                ey[n] = 10 * sqrt(y[n]);
+        Int_t nAdd = 0;
+        bool doHitCorrection = true;
+        // bool doHitCorrection = nHits <= 18; //in case we want to apply it only to the smaller events
+        if (doHitCorrection) {
+            nAdd = 2;
+        }
+        Int_t nElems = nHits + nAdd;
+        vector<Double_t> x(nElems), y(nElems), ex(nElems), ey(nElems);
+        Int_t k = nAdd / 2;
+        Double_t xMin = std::numeric_limits<double>::max();
+        Double_t xMax = std::numeric_limits<double>::lowest();
+        for (int n = 0; n < GetNumberOfHits(); k++, n++) {
+            x[k] = fX[n];
+            y[k] = fEnergy[n];
+            ex[k] = 0;
+            xMin = min(xMin, x[k]);
+            xMax = max(xMax, x[k]);
+            if (y[k] != 0) {
+                ey[k] = 10 * sqrt(y[k]);
             } else {
-                ey[n] = 0;
+                ey[k] = 0;
             }
         }
-        TGraphErrors* grX = new TGraphErrors(nHits, &x[0], &y[0], &ex[0], &ey[0]);
-        Double_t maxY = MaxElement(nHits, grX->GetY());
-        Double_t maxX = grX->GetX()[LocMax(nHits, grX->GetY())];
+        Int_t h = nHits + nAdd / 2;
+        if (doHitCorrection) {
+            x[0] = xMin - 0.5;
+            x[h] = xMax + 0.5;
+            y[0] = 0.0;
+            y[h] = 0.0;
+            ex[0] = 0.0;
+            ex[h] = 0.0;
+            ey[0] = 70.0;
+            ey[h] = 70.0;
+        }
+        TGraphErrors* grX = new TGraphErrors(nElems, &x[0], &y[0], &ex[0], &ey[0]);
+        // TCanvas *c = new TCanvas("c","X position fit",200,10,500,500);
+        // grX->Draw();
+        // Defining the starting parameters for the fit.
+        Double_t maxY = MaxElement(nElems, grX->GetY());
+        Double_t maxX = grX->GetX()[LocMax(nElems, grX->GetY())];
+        Double_t sigma = abs(x[0] - x[h]) / 2.0;
+        // std::cout << "maxX: " << maxX << ", maxY: " << maxY << ", sigma: " << sigma << std::endl;
 
         TF1* fit = new TF1("", "gaus");
         fit->SetParameter(0, maxY);
         fit->SetParameter(1, maxX);
-        fit->SetParameter(2, 2.0);
+        fit->SetParameter(2, sigma);
         grX->Fit(fit, "QNB");  // Q = quiet, no info in screen; N = no plot; B = no automatic start
-                               // parmaeters; R = Use the Range specified in the function range
+                               // parameters; R = Use the Range specified in the function range
 
         gausSigmaX = fit->GetParameter(2);
     }
 
-    return gausSigmaX;
+    return abs(gausSigmaX);
```

Comment:
I suggest to add some checks if the fit result is negative or too large. In these cases we can return -1 to represent internal error.

Reply:
I don't think a negative value is an error. It is a valid value. See https://root.cern.ch/doc/master/TMath_8cxx_source.html#l00448

For values that are too large, I'm trying the code below but it doesn't compile and I don't know why. Any suggestion?

```C++
        TF1* fit = new TF1("", "gaus");
        fit->SetParameter(0, maxY);
        fit->SetParameter(1, maxX);
        fit->SetParameter(2, sigma);
        TFitResultPtr fitResult = grX->Fit(fit, "QNBS");  // Q = quiet, no info in screen; N = no plot; B = no automatic start
                               // parameters; R = Use the Range specified in the function range; S = save and return the fit result.
        if(fitResult->isValid()){
            gausSigmaX = fit->GetParameter(2);
        }
        else{
            return -1.0; // the fit failed, return -1 to indicate failure
        }
```
The error I get is the following:

```
Consolidate compiler generated dependencies of target RestFramework
[  1%] Building CXX object source/framework/CMakeFiles/RestFramework.dir/core/src/TRestHits.cxx.o
/home/cristina/GitHub/rest/source/framework/core/src/TRestHits.cxx: In member function ‘Double_t TRestHits::GetGaussSigmaX()’:
/home/cristina/GitHub/rest/source/framework/core/src/TRestHits.cxx:825:21: error: invalid use of incomplete type ‘class TFitResult’
         if(fitResult->isValid()){
                     ^~
In file included from /programas/root/6.24.02/include/TGraph.h:30,
                 from /programas/root/6.24.02/include/TGraphErrors.h:24,
                 from /home/cristina/GitHub/rest/source/framework/core/inc/TRestHits.h:30,
                 from /home/cristina/GitHub/rest/source/framework/core/src/TRestHits.cxx:76:
/programas/root/6.24.02/include/TFitResultPtr.h:30:7: note: forward declaration of ‘class TFitResult’
 class TFitResult;
       ^~~~~~~~~~
make[2]: *** [source/framework/CMakeFiles/RestFramework.dir/build.make:734: source/framework/CMakeFiles/RestFramework.dir/core/src/TRestHits.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:240: source/framework/CMakeFiles/RestFramework.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
cristina@sultan:~/GitHub/rest/build$

```

Reply:
Anyway it is just an suggestion. If you don't return -1 there is no problem for me.